_posts/drafts/hack-go-type-system.md

```diff
@@ -0,0 +1,842 @@
+---
+published: false
+title: Hacking Go's type system
+layout: post
+---
+
+Are you in the mood for a stroll inside Go's type system ?
+If you are alredy familiarized with it, this post can be funny
+for you, or just plain stupid.
+
+If you have no idea how types and interfaces are implemented on Go,
+you may learn something, I sure did :-)
+
+<!-- more -->
+
+Since I worked with handwritten type systems in C,
+like the one found in [glib GObjects](https://developer.gnome.org/gobject/stable/),
+I'm always curious on how languages implement the
+concept of type safety on a machine that actually only
+has numbers. This curiosity has extended to how I could
+bend Go's type system to my own will.
+
+Go's instances do not carry type information on the instances,
+so my only chance will involve using a interface{}. All type systems
+that I'm aware off usually implement types as some sort of integer code,
+which is used to check if the type corresponds to the one you are casting.
+
+To begin the exploration I wanted to find how type assertions are
+made, so I wrote this ridiculous code:
+
+```
+package main
+
+import "fmt"
+
+func main() {
+	var a int
+	var b interface{} = a
+	c := b.(int)
+	fmt.Println(c)
+}
+```
+
+Compiled it outputing it's assembly:
+
+```
+go build -gcflags -S cast.go
```

Comment:
Did not know about that, this post was my first contact with Golang assembly :-). Thanks ;-)